google/resource_compute_instance_from_template.go

```diff
@@ -0,0 +1,136 @@
+	return &schema.Resource{
+		Create: resourceComputeInstanceFromTemplateCreate,
+		Read:   resourceComputeInstanceRead,
+		Update: resourceComputeInstanceUpdate,
+		Delete: resourceComputeInstanceDelete,
+
+		// Import doesn't really make sense, because you could just import
+		// as a google_compute_instance.
+
+		Timeouts: resourceComputeInstance().Timeouts,
+
+		Schema:        computeInstanceFromTemplateSchema(),
+		CustomizeDiff: resourceComputeInstance().CustomizeDiff,
+	}
+}
+
+func computeInstanceFromTemplateSchema() map[string]*schema.Schema {
+	s := resourceComputeInstance().Schema
+
+	for _, field := range []string{"boot_disk", "machine_type", "network_interface"} {
+		s[field].Required = false
+	}
+
+	recurseOnSchema(s, func(field *schema.Schema) {
+		// We don't want to accidentally use default values to override the instance
+		// template, so remove defaults.
+		field.Default = nil
+
+		// Make non-required fields computed since they'll be set by the template.
+		// Leave deprecated and removed fields alone because we don't set them.
+		if !field.Required && !(field.Deprecated != "" || field.Removed != "") {
+			field.Computed = true
+		}
+	})
+
+	s["source_instance_template"] = &schema.Schema{
+		Type:     schema.TypeString,
+		Required: true,
+		ForceNew: true,
+	}
+
+	return s
+}
+
+func recurseOnSchema(s map[string]*schema.Schema, f func(*schema.Schema)) {
+	for _, field := range s {
+		f(field)
+		if e := field.Elem; e != nil {
+			if r, ok := e.(*schema.Resource); ok {
+				recurseOnSchema(r.Schema, f)
+			}
+		}
+	}
+}
+
+func resourceComputeInstanceFromTemplateCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	// Get the zone
+	z, err := getZone(d, config)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] Loading zone: %s", z)
+	zone, err := config.clientCompute.Zones.Get(project, z).Do()
+	if err != nil {
+		return fmt.Errorf("Error loading zone '%s': %s", z, err)
+	}
+
+	instance, err := computeInstance(project, zone, d, config)
+	if err != nil {
+		return err
+	}
+
+	// Force send all top-level fields in case they're overridden to zero values.
+	// TODO: consider doing so for nested fields as well.
+	for f, s := range computeInstanceFromTemplateSchema() {
+		// It seems that GetOkExists always returns true for sets.
+		// TODO: confirm this and file issue against Terraform core.
+		// In the meantime, don't force send sets.
+		if s.Type == schema.TypeSet {
+			continue
+		}
+
+		if _, exists := d.GetOkExists(f); exists {
+			instance.ForceSendFields = append(instance.ForceSendFields, strcase.UpperCamelCase(f))
+		}
+	}
+
+	// Read create timeout
+	// Until "create_timeout" is removed, use that timeout if set.
+	createTimeout := int(d.Timeout(schema.TimeoutCreate).Minutes())
+	if v, ok := d.GetOk("create_timeout"); ok && v != 4 {
+		createTimeout = v.(int)
+	}
```

Comment:
Is there a reason to port over this behavior we want to get rid of to a new resource? I think we could probably just not use `create_timeout` here.

Reply:
Done